File: binary-search/binary_search.go
package main

import "fmt"

func main() {
    Array := []int{1, 2, 3, 4, 5}
    fmt.Println(binarySearch(Array, 0, len(Array), 5))
    fmt.Println(binarySearch(Array, 0, len(Array), 6))
}

func binarySearch(array []int, start, end, target int) int {
    if start > end {
        return -1
    }

    mid := (start + end) / 2

    if mid == len(array) {
        return -1
    } else if (array[mid] == target) {
        return mid
    } else if (target < array[mid]) {
        return binarySearch(array, start, mid - 1, target)
    } else if (target > array[mid]){
        return binarySearch(array, mid + 1, end, target)
    }

    return -1
}

File: searching/binary-search.cpp
#include<bits/stdc++.h>
using namespace std;

int binarySearch(int arr[], int start, int end, int target) {
    while (start <= end) {
        int mid = (start + end) / 2;
        
        if (arr[mid] == target)
            return mid;
        else if (arr[mid] < target) {
            return binarySearch(arr, 0, mid - 1, target);
        } else if (arr[mid] > target) {
            return binarySearch(arr, mid + 1, end, target);
        }
    }

    return -1;
}

int main() {
    int arr[] = {1, 2, 3, 4};
    int size = sizeof(arr) / sizeof(arr[0]);
    cout << binarySearch(arr, 0, size, 3) << endl;
    cout << binarySearch(arr, 0, size, 8) << endl;
}
